core/store/orm/orm.go

```diff
@@ -291,41 +295,123 @@ func (orm *ORM) convenientTransaction(callback func(*gorm.DB) error) error {
 }
 
-// SaveJobRun updates UpdatedAt for a JobRun and saves it
+// SaveJobRun updates UpdatedAt for a JobRun and updates its status, finished
+// at and run results.
+// It auto-inserts run results if none are existing already for the given
+// jobrun/taskruns.
 func (orm *ORM) SaveJobRun(run *models.JobRun) error {
+	if run.ID == uuid.Nil {
+		return errors.New("job run did not have an ID set")
+	}
 	if err := orm.MustEnsureAdvisoryLock(); err != nil {
 		return err
 	}
-	// FIXME: Be neater about this context timeout
-	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	ctx, cancel := postgres.DefaultQueryCtx()
 	defer cancel()
-	return postgres.GormTransaction(ctx, orm.DB, func(dbtx *gorm.DB) error {
-		result := dbtx.Unscoped().
-			//Session(&gorm.Session{FullSaveAssociations: true}). // We want to save the RunResult and TaskRuns also.
-			Where("updated_at = ?", run.UpdatedAt).
-			Omit("deleted_at").
-			Save(run)
+	err := postgres.GormTransaction(ctx, orm.DB, func(dbtx *gorm.DB) error {
+		result := dbtx.Exec(`
+UPDATE job_runs SET "status"=?, "finished_at"=?, "updated_at"=NOW(), "creation_height"=?, "observed_height"=?, "payment"=?
+WHERE updated_at = ? AND "id" = ?`,
+			run.Status, run.FinishedAt, run.CreationHeight, run.ObservedHeight, run.Payment,
+			run.UpdatedAt, run.ID,
+		)
 		if result.Error != nil {
-			if strings.Contains(result.Error.Error(), "duplicate key value violates unique constraint") {
-				return ErrOptimisticUpdateConflict
-			}
-			return result.Error
+			return errors.Wrap(result.Error, "failed to update job run")
 		}
 		if result.RowsAffected == 0 {
 			return ErrOptimisticUpdateConflict
 		}
-		err := dbtx.Unscoped().Save(&run.Result).Error
-		if err != nil {
-			return err
+
+		uRunResultStrs := []string{}
+		uRunResultArgs := []interface{}{}
+		if run.Result.ID > 0 {
+			uRunResultStrs = append(uRunResultStrs, "(?::bigint, ?::jsonb, ?::text)")
+			uRunResultArgs = append(uRunResultArgs, run.ResultID, run.Result.Data, run.Result.ErrorMessage)
+		} else {
+			if run.ResultID.Valid {
+				return errors.Errorf("got valid ResultID %v for job run %s but Result.ID was 0", run.ResultID.Int64, run.ID)
+			}
+			// Insert the run result for the job run
+			err := dbtx.Exec(`
+WITH run_result AS (
+	INSERT INTO run_results (data, error_message, created_at, updated_at) VALUES(?, ?, NOW(), NOW()) RETURNING id
+)
+UPDATE job_runs
+SET result_id = run_result.id
+FROM run_result
+WHERE job_runs.id = ?
+`, run.Result.Data, run.Result.ErrorMessage, run.ID).Error
+			if err != nil {
+				return errors.Wrap(err, "error inserting run result for job_run")
+			}
 		}
-		if len(run.TaskRuns) > 0 {
-			err = dbtx.Unscoped().Save(&run.TaskRuns).Error
+
+		uTaskRunStrs := []string{}
+		uTaskRunArgs := []interface{}{}
+		for _, tr := range run.TaskRuns {
+			uTaskRunStrs = append(uTaskRunStrs, "(?::uuid, ?::run_status)")
+			uTaskRunArgs = append(uTaskRunArgs, tr.ID, tr.Status)
+			if tr.Result.ID > 0 {
```

Comment:
same here - are guaranteed to have the Result preloaded? Might be good to have a defensive check anyways

Reply:
The defensive check is on line 358 below

Reply:
I added tests for both of these cases